processor/src/debug.rs

```diff
@@ -45,8 +47,46 @@ impl VmStateIterator {
             process,
             error: result.err(),
             clk: 0,
+            asmop_idx: 0,
         }
     }
+
+    /// Returns the asm op state corresponding to this vm state and whether this is the start of
+    /// operation sequence corresponding to current assembly instruction.
+    fn get_asmop(&self) -> (Option<AsmOpState>, bool) {
+        let mut asmop: Option<AsmOpState> = None;
+        let assembly_ops = self.process.decoder.assembly_ops();
+        if self.clk != 0 && self.asmop_idx <= assembly_ops.len() {
+            let next_asmop = if self.asmop_idx < assembly_ops.len() {
+                &assembly_ops[self.asmop_idx]
+            } else {
+                &assembly_ops[self.asmop_idx - 1]
+            };
+            let (curr_asmop, op_idx) = if self.asmop_idx > 0 {
+                (
+                    &assembly_ops[self.asmop_idx - 1],
+                    (self.clk - assembly_ops[self.asmop_idx - 1].0) as u8,
+                )
+            } else {
+                (next_asmop, 0) //dummy value, never used.
+            };
+            if next_asmop.0 == self.clk - 1 {
+                asmop = Some(AsmOpState::new(
+                    next_asmop.1.op().clone(),
+                    next_asmop.1.num_cycles(),
+                    1,
+                ));
+                return (asmop, true);
+            } else if self.asmop_idx > 0 && op_idx <= curr_asmop.1.num_cycles() {
+                asmop = Some(AsmOpState::new(
+                    curr_asmop.1.op().clone(),
+                    curr_asmop.1.num_cycles(),
+                    op_idx,
+                ));
+            }
+        }
+        (asmop, false)
```

Comment:
I find the logic here a bit difficult to follow. Maybe break this up into simpler conditions (it's OK if this results in more code). For example, the begging could be something like this:

```Rust
if self.asmop_idx == assembly_ops.len() {
    return (None, false);
}
```



Also, probably adding a few internal comments would be good.

Reply:
Please let me know if it's clear now.